### PR TITLE
`sparsify.apply`

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR represents the feature branch for `sparsify.apply` cli push

### Background:
sparsify.apply is a CLI utility, used to run the specified experiments initialized using `sparsify.init`

```bash
sparsify.apply --help
Usage: sparsify.apply [OPTIONS]

  sparsify.apply CLI utility to apply sparsify to a given experiment, or
  initialize and run an experiment if not created yet.

Options:
  --experiment-type [sparse-transfer|one-shot|training-aware]
                                  The type of the experiment to run.
  --use-case [document_classification|image_classification|information_retrieval|masked_language_modeling|multilabel_text_classification|object_detection|question_answering|segmentation|sentiment_analysis|text_classification|token_classification]
                                  The task this model is for.
  --project-id TEXT               Id of the project this run belongs to.
  --experiment-id TEXT            Id of the experiment this run belongs to.
  --working-dir TEXT              Path to save the deployment ready model to.
                                  [default:
                                  /home/rahul/github_projects/sparsify]
  --model TEXT                    Path to model.
  --teacher TEXT
  --optimizer [Adadelta|Adagrad|Adam|AdamW|SparseAdam|Adamax|ASGD|SGD|RAdam|Rprop|RMSprop|NAdam|LBFGS]
                                  The optimizer to use
  --recipe TEXT                   Recipe to override automatic recipe.
  --recipe-args TEXT
  --data TEXT                     Path to dataset folder containing training
                                  data and optionally validation data.
  --eval-metric [kl|accuracy|mAP|recall|f1]
                                  Metric that the model is evaluated against
                                  on the task.  [default: kl]
  --train-samples INTEGER         Number of train samples to use from the
                                  dataset for processing. Will use all train
                                  samples if not specified.
  --val-samples INTEGER           Number of validation samples to use from the
                                  dataset for processing. Will use all eval
                                  samples if not specified.
  --deploy-hardware TEXT
  --deploy-engine [deepsparse|onnxruntime]
                                  [default: deepsparse]
  --deploy-scenario TEXT
  --optim-level FLOAT             Preferred tradeoff between accuracy and
                                  performance. Float value in the range [0,
                                  1]. Default 0.5  [default: 0.5]
  --model-id TEXT                 sparsify model id.
  --version                       Show the version and exit.  [default: False]
  --help                          Show this message and exit.  [default:
                                  False]
```

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204285970834858

### CODE REVIEW

File: src/sparsify/apply.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    """
+    sparsify.apply CLI utility to apply sparsify to a given experiment, or initialize
+    and run an experiment if not created yet.
+    """
+    set_log_level(logger=_LOGGER, level=logging.DEBUG if debug else logging.INFO)
+    apply(**kwargs)
+    _LOGGER.debug(f"locals: {locals()}")
+
+
+def apply(**kwargs):
```
Comment (rahul-tuli) [inline src/sparsify/apply.py:117]: I do not like the usage of kwargs here, but the number of arguments quickly blew up to too many, I would like to convert the kwargs to a bunch of typed objects, as specified in the product doc in a future PR, keeping this as is for now

File: src/sparsify/apply.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        )
+        kwargs.update({"experiment_type": experiment_type})
+
+    if kwargs["project_id"] is None:
+        project_id = client.get(url=f"/experiments/{experiment_id}/project_id")
+        kwargs.update({"project_id": project_id})
+
+    command_args = []
+    if experiment_type == "sparse-transfer":
+        command_args = [
```
Comment (rahul-tuli) [inline src/sparsify/apply.py:142]: The typed objects, could provide an easier way to generate a command string in the future

File: src/sparsify/apply.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            project_id=project_id,
+            experiment_id=experiment_id,
+            analysis_type="model_analysis",
+            analysis_file=analysis_file_path,
+        )
+
+        client.update_experiment_status(
+            experiment_id=experiment_id, status=ExperimentStatus.COMPLETE.value
+        )
+    except Exception as exception:
```
Comment (rahul-tuli) [inline src/sparsify/apply.py:201]: Would also like to narrow this down to specific exceptions, but need to look deeper into all kinds of exceptions that could be raised either by subprocess or by `sparsify.run` utility

File: src/sparsify/apply.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+@opts.add_deploy_opts
+@opts.OPTIM_LEVEL
+@click.option("--model-id", help="sparsify model id.")
+@click.version_option(version=version_major_minor)
+@click.option("--max-latency", type=float, help="Maximum latency in ms.")
+@click.option(
+    "--min-throughput", type=float, help="Minimum throughput in images per second."
+)
+@click.option("--max-size", type=float, help="Maximum model size in MB.")
+@click.option("--max-memory", type=float, help="Maximum memory usage in MB.")
```
Comment (bfineran) [inline src/sparsify/apply.py:102]: these args have been removed from the product (doc wasn't updated)

File: src/sparsify/apply.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        for key, value in kwargs.items():
+            if value is not None:
+                command_args.extend([f"--{key}", value])
+    else:
+        raise ValueError(f"Invalid experiment type {experiment_type}")
+
+    command_str = ["sparsify", "run", experiment_type].extend(command_args)
+
+    try:
+        client.update_experiment_status(
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparsify/apply.py:178]: Agree!

File: src/sparsify/apply.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            url=f"/experiments/{experiment_id}/experiment_type"
+        )
+        kwargs.update({"experiment_type": experiment_type})
+
+    if kwargs["project_id"] is None:
+        project_id = client.get(url=f"/experiments/{experiment_id}/project_id")
+        kwargs.update({"project_id": project_id})
+
+    command_args = []
+    if experiment_type == "sparse-transfer":
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparsify/apply.py:131]: Resolving this, it was discussed offline to keep this as is for this PR, this code will be revamped when Pydantic-Click support lands

**Review by KSGulin — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**